content/topics/java-specific/oop-basics/_index.md

```diff
@@ -1,15 +1,179 @@
+
+### Inheritance
+This OOP principle allows one class to inherit all the attributes and behaviors of another class.
+Inheritance has a hierarchy  relationship between superclasses (base classes )and subclasses
+
+The following Java code shows how Inheritance can be implemented:
+```
+//superclass
+public class Vehicle {
+    void printType(){
+        System.out.println("I am a Vehicle");
+    }
+}
+
+public class Car extends  Vehicle {
+    //Override method 
+    @Override
+    void printType() {
+        //call method in super class
+        super.printType();
+        System.out.println("I am a car");
+    }
+
+
+    public static void main(String[] args) {
+        //Create a car object
+        Car car = new Car();
+        //call method
+        car.printType();
+    }
+
+}
+
+```
+
+### Polymorphism
+Polymorphism in Java is a concept by which we can perform a single action in different ways. Polymorphism is derived from 2 Greek words: poly and morphs. The word "poly" means many and "morphs" means forms. So polymorphism means many forms.
+
+There are two types of polymorphism in Java: compile-time polymorphism and run-time polymorphism. We can perform polymorphism in java by method overloading and method overriding.
+#### Method overriding
+
+Overriding a method is when a method in the subclass has the same name and method signature as a method in the superclass. When overriding a method you are not allowed to make the method more private. 
+Example:
+You cannot make a public method protected.
+
+The following Java code shows how overriding can be implemented:
+```
+public class Animal {
+    public void eat() {
+        System.out.println("Generic animal is eating");
+    }
+    static class Horse extends Animal {
+
+        //Override method
+        @Override
+        public void eat() {
+            System.out.println("Horse is eating");
+        }
+
+        public static void main(String[] args) {
+            //Create a car object
+            Horse horse = new Horse();
+            //call method
+            horse.eat();
+        }
+    }
+}
+
+```
+#### Method overloading
+
+Overloading a method is when a method in the subclass has the same name but the  method signature is different from the method in the superclass.
+
+The following Java code shows how overloading can be implemented:
+```
+public class Calculate{
+
+    public int product (int x, int y) {
+        return (x * y);
+    }
+
+    // Overloaded. This product method  takes three int parameters
+    public int product(int x, int y, int z) {
+        return (x * y * z);
+    }
+
+    // Overloaded. This product method takes two double parameters
+    public double product(double x, double y) {
+        return (x * y);
+    }
+
+    public static void main(String args[]) {
+       Calculate prod_object = new Calculate();
```

Comment:
again. Rather put the main somewhere else.

Another way to think about it. Imagine we want to write reusable code (stranger things have happened). Now we ubild this Calculator and we incorperate it into another program. Then the other program freaks out because it has two main methods. Maybe three main members because we are calculating some stuff about Horses. 

Your code should be reusabe without having to be re-written. A class should only do the sstuff that it is responsible for.